### ISSUE TITLE
make Elektra completely introspectable

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#3888]
While it is possible to get/set all keys and specifications of an application, it is currently not possible to determine which parentKey an application actually is using.

To fix this problem, I propose that:

1. applications add some static information in their binaries that tells that they use Elektra, gopts and which parentKey. This could be in some ELF section.
2. gopts always accepts some command-line options like `--elektra-help` and `--elektra-print-parent`.

Option 1. alone is not enough as some applications might use profiles and end up with different parent keys in some situations.

Option 2. alone is not enough as some applications might successfully do something even if `--elektra-print-parent` is passed to them (in a quick test e.g. latex).

@kodebach @dev2718 What do you think?

**Comment by markus2330** [comment]
> Is this common? Do you have an example of another library that does this?

`strings` is a common tool but it is probably not so common to embed something in an executable just for that reason. Within companies it is probably more common to add build information (like version information) to executables.

I am afraid environment variables have the exact same problem as command-line options: they will not work on non-elektrified applications.

I think our first goal should be to find a reliable way to determine if Elektra and gopts is being used, and then we can find a communication protocol to the executable to get dynamic data. (If it is needed at all, maybe we simply provide a library routine for the profiles and then anybody can recalculate to get the same results).

**Comment by dev2718** [comment]
What about we reserve something on the lines of `system:/elektra/introspection/*` as a virtual/in memory section of the kdb?
Then, whenever gopts (or any other plugin) is called by some process, gopts can report introspection data, e.g. `system:/elektra/introspection/<pid>/gopts/parentKey`, that lives as long as the process exists.

**Comment by kodebach** [comment]
> What about we reserve something on the lines of `system:/elektra/introspection/*` as a virtual/in memory section of the kdb?

How would you access that from outside the process, if it is in-memory only? We could instead write it to disk and read it via the FUSE binding, but that would mean modifying the KDB during `kdbGet()`, which is something I'd like to avoid. There would also be a problem, if a process crashes. We'd have some data in `system:/elektra/introspection/<pid>` that is no longer correct and wouldn't be removed.



> I am afraid environment variables have the exact same problem as command-line options: they will not work on non-elektrified applications.

If an application is not elektrified, what is there to introspect? The advantage of environment variables is every application I know of ignores unknown environment variables (and for good reason), while many applications refuse unknown command-line options and just don't work with them.

> I think our first goal should be to find a reliable way to determine if Elektra and gopts is being used, and then we can find a communication protocol to the executable to get dynamic data. 

**TL;DR:** Without executing the application and watching for something that happens within or immediately before e.g. `kdbGet()` there is no 100% reliable way to say whether `kdbGet()` would be called (same for `gopts`). This is just a fact of Turing completeness.

I'll assume the application is dynamically linked. For statically linked applications see the note at the bottom.

Detecting whether Elektra could be used is easy:

```sh
# use e.g. bin/gopts as myapp (when in the build folder for Elektra)
ldd myapp | grep libelektra-kdb.so
```

That doesn't really tell us whether `kdbGet()` is actually ever called, but without executing the application we'll never know that for certain.

Since `gopts` is loaded via `dlopen`, we can see that from just looking at the ELF metadata. We have to execute the application. For that we can use `LD_DEBUG`:

```sh
LD_DEBUG=libs myapp --help 2>&1 | grep "calling init: .*/libelektra-gopts.so"
```

> Note: `libelekra-opts` and the `gopts` plugin always support `--help` and `--help` cannot be used by the spec. So if the application is implemented correctly (i.e. it correctly checks the value of `proc:/elektra/gopts/help` and exits when it is `1`), then the application shouldn't really do anything except printing the help message.

However, I fail to see how that would be any different than e.g.

```sh
ELEKTRA_INTROSPECT=kdbGet myapp --help 2>&1 | grep "kdbGet() called"
```

In both solutions `grep` just wouldn't find a match, if Elektra/`gopts` isn't used.

You could also use `ltrace`, which would give you more infos than `ldd` and `LD_DEBUG`, but less infos than a custom solution could.

If we don't want to bake these introspection routines into `libelektra-kdb` (for size and maybe security reasons), we could create as separate `libelektra-kdb-introspect.so` that can be loaded via `LD_PRELOAD`. It would provide versions of `kdbGet()` et al. that wrap the original `kdbGet()` et al. and do the introspections. This could then be called like this:

```sh
# ELEKTRA_INTROSPECT could be made optional
LD_PRELOAD=libelektra-kdb-introspect.so ELEKTRA_INTROSPECT=kdbGet myapp --help 2>&1 | grep "kdbGet() called"
```

Another way would be to build the introspection points into `libelektra-kdb`, but make them optionally implemented. For example, `kdbGet()` et al. would call `elektraIntrospectLog (...)` at various points. But `elektraIntrospectLog (...)` is either an empty stub that can be replaced via `LD_PRELOAD` or it used `dlsym` to see if an actual implementation was loaded. The advantage here is that this would provide a unified and stable API that could safely be called from plugins.

The `LD_PRELOAD` solutions, could also come with an `__attribute__((destructor))` function that could report "Elektra not used", if no introspection call was made.

> If it is needed at all, maybe we simply provide a library routine for the profiles and then anybody can recalculate to get the same results

Not sure what you mean with that, but if you want to call a function exported by the application than that would require support from the application developer (or at least the person who compiled the application). If you mean, we provide a function in `libelektra-kdb.so` (or a command for the `kdb` tool) then we would still need to know the arguments that the application passed to e.g. `kdbGet()`.

You're only thinking about profiles (which are resolved by Elektra), but the application could have some custom logic (not encouraged but possible) or could use multiple KDB instances (e.g. multiple independent threads). In both of these cases, I don't see a reliable solution without building _something_ into `kdbGet()`.

---

For statically linked applications, I think only a custom solution baked directly into the static version of Elektra would work reliably. We could add ELF sections, but is building statically linked applications, they're likely on an embedded device and have binary size constraints. In that case they probably strip out unnecessary ELF sections. But in any case, I don't think we need to support statically linked applications for this, because the people using those probably know what their applications do and how to debug them.

---

@markus2330 Please clarify what exactly you expect from this introspection. Should it work without executing the application? If so, what information should be available without execution?

**Comment by markus2330** [comment]
> system:/elektra/introspection/*

From the user's perspective this is a good idea if we can guarantee up-to-date and correct information there.

> If an application is not elektrified, what is there to introspect?

Nothing but how can an application/tool know that? We need know for which processes it is sensible to offer the introspection of `proc:`.

> [a few methods how to find out kdbGet gets invoked]

I don't think we need to know if kdbGet actually gets invoked. It is about the intention of the app-developer, who wants to offer a fully elektrified application and wants the ability to introspect the app. So this is actually a static information.

> then we would still need to know the arguments that the application passed to e.g. kdbGet().

Exactly this is what this proposal is about: how can we ask binaries which arguments it (may) pass to kdbGet? (Which currently boils down to the parentKey.)

>You're only thinking about profiles (which are resolved by Elektra), but the application could have some custom logic (not encouraged but possible) or could use multiple KDB instances (e.g. multiple independent threads).

Yes, in the long term we need to clarify how to:

1. dynamically get parentKeys (profiles or other logic)
2. support multiple KDB instances
3. support binaries that are actually many different tools (like git)

> @markus2330 Please clarify what exactly you expect from this introspection. Should it work without executing the application?

Executing the application would be fine if we find a way to

1. know that it is safe to execute, or
2. safely execute non-elektrifed applications.

> If so, what information should be available without execution?

As first version a static parentKey per executable should do.

**Comment by kodebach** [comment]
Okay I think we need more clarification. What purpose will this kind of introspection serve?

My understanding was that a user/admin wants to understand what config is loaded by an application. They run the application just like they normally would but enable the introspection. This is why I suggested the approach similar to `LD_DEBUG`, since that fulfils a similar purpose. I want to know what dynamic libraries `myapp` loads, so I run `LD_DEBUG=libs myapp`.

I think you want a way for e.g. the FUSE binding to avoid the `--parentKey` option. And you want the solution to work for already running processes.

> I don't think we need to know if kdbGet actually gets invoked.

If `kdbGet()` isn't called, from an introspection POV the application is essentially the same as a non-elektrified one. If it doesn't call `kdbGet()` it doesn't load any data from the KDB, so the `parentKey` is irrelevant.

> As first version a static parentKey per executable should do.

Again, how is this better than just putting the parentKey into the documentation? Also static information is not really introspection.

> if we can guarantee up-to-date and correct information

Like I said, the only way to guarantee up-to-date and correct information is to capture `kdb*` calls as they happen. If we try to reproduce a `kdbGet()` call a lot of things could be different: storage files changed, plugin is non-deterministic, even the `.so` file could have changed.

**Comment by markus2330** [comment]
> Okay I think we need more clarification. What purpose will this kind of introspection serve?

To clarify: I am talking about introspection of configs and specs, not about what the application is doing.

The main part of introspection we already have: as processes are encouraged to write all conf changes via kdbSet (this is also the reason why we do not have kdbValidate), tools can introspect with kdbGet the configs of tools if:

1. you know the parent Key
2. you have the same chroot (system:)
3. you have the same user (user:)
4. you have the same current working directory (dir:)
5. you know the command-line flags & env vars (proc:)

This issue is only about 1., for the other parts we already found solutions (for 2. it is not implemented, though). 1. was first to believed to be not such a big issue, as we thought it is enough if people look at the documentation. As we saw in FUSE, however, the parentKey is also required for to introspect proc:, so an automatic solution to find 1. for an executable (or even running process) is needed.

> I think you want a way for e.g. the FUSE binding to avoid the --parentKey option. And you want the solution to work for already running processes.

Exactly, although the requirement to "look up docu to find out the parentKey" also was disturbing. So what we need is: given an executable, tell me which parentKey(s) it uses.

Note that it would be enough if can be answered for executables as it is always possible to execute /proc/self/exe again.

> If kdbGet() isn't called, from an introspection POV the application is essentially the same as a non-elektrified one. If it doesn't call kdbGet() it doesn't load any data from the KDB, so the parentKey is irrelevant.

The application might call `kdbGet` at any later point.

> Again, how is this better than just putting the parentKey into the documentation?

That tools cannot get it programmatically.

> Also static information is not really introspection.

To keep the implementation sane, we'll also put some requirements to the applications:

- doing kdbSet when some internal config (state) has changed
- doing kdbGet when some external config has changed (notification)
- only use one parentKey per handle (or only keys below for subsequent kdbGet/kdbSet calls)
- respect a protocol how to retrieve the parentKey(s)

With these requirements, we will hopefully be able that the "static information" retrieved by kdbGet and the executable is exactly what the running processes are actually using.

**Comment by kodebach** [comment]
> tools can introspect with kdbGet the configs of tools if: [...]

You missed:

6. you used the same contract when calling `kdbOpen()`
7. you have the same data in the KeySet passed to `kdbGet()` (normally empty, but if not it could have an effect)
8. all plugins in relevant mountpoints are deterministic, i.e. their output is 100% predictable based on the other conditions

> The application might call `kdbGet` at any later point.

IMO "introspection" means: Inspecting the state of a system while it runs, i.e. see what actually happend not what might happen in some other cases.

> The main part of introspection we already have: as processes are encouraged to write all conf changes via kdbSet

This does not help with introspection. There could still have been an external change directly to the file.

>  As we saw in FUSE, however, the parentKey is also required for to introspect proc:, so an automatic solution to find 1. for an executable (or even running process) is needed.

For FUSE we already have another (IMO much better) solution. Additionally, we can never provide an automatic static solution _without support from the application developer_ (more specifically the person who compiles the application). We don't know the parent key when compiling Elektra, the only thing we could do is wait for `kdbGet()` to be called and then do something.

> To keep the implementation sane, we'll also put some requirements to the applications:

IMO it would be much saner to just have a wrapper that simply logs the input and output of `kdb*` calls somewhere. There is very little that could go wrong, it is not very hard to implement (definitely not harder than custom ELF sections) and doesn't put any requirements on application developers. Especially not on how they compile their application.

All these informal unchecked requirements and specifications are just asking for problems...

**Comment by markus2330** [comment]
There is even more:

9. The value returned by ksLookup("/...") is used as-is and as specified (without transformation in the code)
10. No external changes to files (only using Elektra's tooling)

Some of these requirements can be taken care of in high-level APIs. But at some point devs also must have interest in participating properly. Imho it is unrealistic to aim for solutions where application developers don't need to consider anything.

> IMO it would be much saner to just have a wrapper that simply logs the input and output of kdb* calls somewhere.

Yes, I like the idea, the "logging" could be to /run/elektra/<pid>. There we could also serialize the initial keyset and the contract, which would greatly reduce the requirements. This could be done by a global plugin mounted by default.
